File: app/models/support_ticket.py
from datetime import datetime, timezone
from enum import Enum
from typing import Optional
from beanie import Document, Indexed
from pydantic import Field


class SupportTicketStatus(str, Enum):
    PENDING = "pending"
    SOLVED = "solved"


class SupportTicket(Document):
    """MongoDB Document model for Customer Support Reports & Tickets."""

    ticket_id: Indexed(str, unique=True)
    user_id: Optional[str] = None
    name: str
    email: str
    phone: str
    subject: str
    details: str
    status: SupportTicketStatus = SupportTicketStatus.PENDING
    admin_notes: Optional[str] = None
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    resolved_at: Optional[datetime] = None

    class Settings:
        name = "support_tickets"
        indexes = [
            "status",
            "email",
            "phone",
            "user_id",
            [("status", 1), ("created_at", -1)],
        ]

    def touch(self):
        """Update updated timestamp."""
        pass

File: src/models/SupportTicket.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum SupportTicketStatus {
  PENDING = 'pending',
  SOLVED = 'solved',
}

export interface ISupportTicket extends Document {
  ticket_id: string;
  user_id?: string;
  name: string;
  email: string;
  phone: string;
  subject: string;
  details: string;
  status: SupportTicketStatus;
  admin_notes?: string;
  created_at: Date;
  resolved_at?: Date;
  touch(): void;
}

const SupportTicketSchema = new Schema<ISupportTicket>(
  {
    ticket_id: { type: String, required: true, unique: true, index: true },
    user_id: { type: String, index: true },
    name: { type: String, required: true },
    email: { type: String, required: true, index: true },
    phone: { type: String, required: true, index: true },
    subject: { type: String, required: true },
    details: { type: String, required: true },
    status: { type: String, enum: Object.values(SupportTicketStatus), default: SupportTicketStatus.PENDING, index: true },
    admin_notes: { type: String },
    created_at: { type: Date, default: Date.now },
    resolved_at: { type: Date },
  }
);

SupportTicketSchema.index({ status: 1, created_at: -1 });

SupportTicketSchema.methods.touch = function () {};

export const SupportTicket = mongoose.model<ISupportTicket>('SupportTicket', SupportTicketSchema, 'support_tickets');
